HW2/P3/mandelbrot.pyx

```diff
@@ -3,45 +3,99 @@ cimport numpy as np
 cimport cython
 import numpy
 cimport AVX
-from cython.parallel import prange
+from cython.parallel import prange, parallel
+from libc.stdlib cimport abort, malloc, free
 
 
-cdef np.float64_t magnitude_squared(np.complex64_t z) nogil:
-    return z.real * z.real + z.imag * z.imag
+# cdef np.float64_t magnitude_squared(np.complex64_t z) nogil:
+#     return z.real * z.real + z.imag * z.imag
 
 @cython.boundscheck(False)
 @cython.wraparound(False)
+
+cdef void print_complex_AVX(AVX.float8 real,
+                             AVX.float8 imag) nogil:
+    cdef:
+        float real_parts[8]
+        float imag_parts[8]
+        int i
+
+    AVX.to_mem(real, &(real_parts[0]))
+    AVX.to_mem(imag, &(imag_parts[0]))
+    with gil:
+        for i in range(8):
+            print("    {}: {}, {}".format(i, real_parts[i], imag_parts[i]))
+
 cpdef mandelbrot(np.complex64_t [:, :] in_coords,
                  np.uint32_t [:, :] out_counts,
                  int max_iterations=511):
     cdef:
-       int i, j, iter
-       np.complex64_t c, z
-
-       # To declare AVX.float8 variables, use:
-       # cdef:
-       #     AVX.float8 v1, v2, v3
-       #
-       # And then, for example, to multiply them
-       #     v3 = AVX.mul(v1, v2)
-       #
-       # You may find the numpy.real() and numpy.imag() fuctions helpful.
+        int i, j, jj, iter
+        np.complex64_t c, z
+        AVX.float8 c_real, c_imag, z_real, z_imag,z_real_new, magnitude_squared, cont_case,avx_out_counts
+        #thread local buffer for storing out counts in memory before populating 
+        #out_counts array
+        float *out_counts_temp
+        int arrsize0=in_coords.shape[0]
+        int arrsize1=in_coords.shape[1]
+        np.float32_t [:,:] in_coords_real=np.zeros((arrsize0,arrsize1),dtype=np.float32)
+        np.float32_t [:,:] in_coords_imag=np.zeros((arrsize0,arrsize1),dtype=np.float32)
+
 
     assert in_coords.shape[1] % 8 == 0, "Input array must have 8N columns"
     assert in_coords.shape[0] == out_counts.shape[0], "Input and output arrays must be the same size"
     assert in_coords.shape[1] == out_counts.shape[1],  "Input and output arrays must be the same size"
-
-    with nogil:
-        for i in range(in_coords.shape[0]):
-            for j in range(in_coords.shape[1]):
-                c = in_coords[i, j]
-                z = 0
+    for i in range(in_coords.shape[0]):
+        for j in range(in_coords.shape[1]):
```

Comment:
You can specify a stepsize in prange to avoid j*8. 
Example of stepsize=8:
 for j in prange(0, in_coordsReal.shape[1], 8):